### PR TITLE
[receiver/receivercreator] Add support for k8s service discovery

### PR DESCRIPTION
**Pull request by a-thaler**
**Description:**
- Added a new watch to the k8s_observer extension for k8s services, which can be enabled using a new flag "observe_services".
- Discovered entities are transformed into a new endpoint type `k8s.service`.
- Adjusted the receivercreator to support the new type `k8s.service`

**Link to tracking Issue:** [#29021](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29021)

**Testing:** Added unit tests analogue to the available tests

**Documentation:** Adjusted readme's of k8s_observer and receivercreator. Added description of new flags and typers.

**Note:**
Current implementation is working as described in the linked ticket. Please check the potential discussion points mentioned in the ticket: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29021#issuecomment-1801290614

### CODE REVIEW

File: extension/observer/k8sobserver/handler.go [hunk 1]
```diff
@@ -76,6 +78,18 @@ func (h *handler) OnUpdate(oldObjectInterface, newObjectInterface interface{}) {
 			newEndpoints[e.ID] = e
 		}
 
+	case *v1.Service:
+		newService, ok := newObjectInterface.(*v1.Service)
+		if !ok {
```
Comment (rmfitzpatrick) [inline extension/observer/k8sobserver/handler.go:83]: I'm not sure why we're not info/warn-level logging these as programming errors. Does it make sense to begin doing so here?

File: extension/observer/k8sobserver/config.go [hunk 2]
```diff
@@ -34,12 +34,14 @@ type Config struct {
 	// it will only discover node endpoints whose `metadata.name` matches the provided node name. If `true` and
 	// Node isn't specified, it will discover all available node endpoints. `false` by default.
 	ObserveNodes bool `mapstructure:"observe_nodes"`
+	// ObserveServices determines whether to report observer service and port endpoints. `false` by default.
+	ObserveServices bool `mapstructure:"observe_services"`
```
Comment (rmfitzpatrick) [inline extension/observer/k8sobserver/config.go:38]: Not needed w/ these changes but I wonder if moving to an inner `observe` map is preferable as we'd add more types.

**Review by rmfitzpatrick — APPROVED** [review]

File: extension/observer/k8sobserver/handler.go (hunk 1 above)
Comment (a-thaler) [reply extension/observer/k8sobserver/handler.go:84]: I added warn statements, please have a look

**Comment by a-thaler** [comment]
Fixed all problems reported by the github actions, requires another run now.

File: extension/observer/endpoints.go [hunk 3]
```diff
@@ -92,6 +95,40 @@ func (e Endpoint) equals(other Endpoint) bool {
...
+	UID string
+	// Labels is a map of user-specified metadata.
+	Labels map[string]string
+	// Annotations is a map of user-specified metadata.
+	Annotations map[string]string
+	// Namespace must be unique for services with same name.
+	Namespace string
+	// ClusterIP is the IP under which the service is reachable within the cluster.
+	ClusterIP string
+	// The type of the service: ClusterIP, NodePort, LoadBalancer, ExternalName
```
Comment (atoulme) [inline extension/observer/endpoints.go:112]:
```suggestion
	// ServiceType is the type of the service: ClusterIP, NodePort, LoadBalancer, ExternalName
```

**Review by atoulme — APPROVED** [review]
just one doc nit

**Pull request merged**